### DISCUSSION TITLE
Feature: Convenience macros for DECLARING to_json and from_json?

### DISCUSSION THREAD

**Discussion by serg06** [nlohmann/json#2369] [Ideas]
I see there is the `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE` macro for defining `to_json` and `from_json`, but there's nothing for just declaring them, is there?

For example if I want to do something like this:

`CustomType.h`:

```
#include "nlohmann/json-fwd.hpp"

struct CustomType
{
	int var1;
	int var2;
};

// DECLARE json conversions
void to_json(nlohmann::json& j, const CustomType& obj);
void from_json(const nlohmann::json& j, CustomType& obj);
```

`CustomType.cpp`:

```
#include "nlohmann/json.hpp"

// DEFINE json conversions
void to_json(nlohmann::json& j, const CustomType& obj)
{
	j["var1"] = obj.var1;
	j["var2"] = obj.var2;
}

void from_json(const nlohmann::json& j, CustomType& obj)
{
	j.at("var1").get_to(obj.var1);
	j.at("var2").get_to(obj.var2);
}
```

I can use the macro to replace the code in `CustomType.cpp`, but I still need the `to_json` and `from_json` declarations in `CustomType.h`.

**Comment by nlohmann** [comment]
There is still no `NLOHMANN_DECLARE_TYPE_*` counterpart to `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE` for declaring (without defining) `to_json`/`from_json` in a header, for use with an out-of-line definition in a .cpp file. This is a reasonable, small, self-contained addition for the common header/source-split pattern, especially now that the DEFINE macros are templated (see the related out-of-line linkage discussion in #4199), and would pair naturally with them.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
